iwyu_include_picker.cc

```diff
@@ -1425,18 +1425,15 @@ void IncludePicker::ExpandRegexes() {
     const string& hdr = incmap.first;
     for (const string& regex_key : filepath_include_map_regex_keys) {
       const vector<MappedInclude>& map_to = filepath_include_map_[regex_key];
-      // Enclose the regex in ^...$ for full match.
-      std::string regex("^" + regex_key.substr(1) + "$");
-      if (RegexMatch(regex_dialect, hdr, regex) &&
+      if (RegexMatch(regex_dialect, hdr, regex_key.substr(1)) &&
           !ContainsQuotedInclude(map_to, hdr)) {
         Extend(&filepath_include_map_[hdr], filepath_include_map_[regex_key]);
         MarkVisibility(&include_visibility_map_, hdr,
                        include_visibility_map_[regex_key]);
       }
     }
     for (const string& regex_key : friend_to_headers_map_regex_keys) {
-      std::string regex("^" + regex_key.substr(1) + "$");
-      if (RegexMatch(regex_dialect, hdr, regex)) {
+      if (RegexMatch(regex_dialect, hdr, regex_key.substr(1))) {
```

Comment:
Since c++17 is the default now, `regex_key` could be a `std::string_view` thus saving the temporary object from `std::string::substr()`. But the `std::regex` interface doesn't support that so there would be still be a copy for that. Same applies above. But because of the anchoring there would still be a copy for the LLVM dialect as well. Still it would save a copy...

Not having `constexpr` or `std::string_view` support in `std::regex` makes me wonder if this was just dumped into the standard and abandoned afterwards since it is unusable anyways. Or maybe it is just too complex to support those.

And you call `std::string::substr()` in both cases. So maybe it could already be stored as that value.

Reply:
Now that that `substr` copy is an rvalue, I believe C++'s copy elision will take care of eliminating temporaries better.

The ultimate goal has to be to act on that `TODO` a few lines up; compile the regexes once in `ExtractKeysMarkedAsRegexes`, and then evaluate them on the include tree separately. But that's only possible once we can store a compiled regex, and that's where surgery like the `PolyRegex` example I posted earlier is necessary.

So I think we can just leave this alone for now, since `std::regex` is opt-in and `llvm::Regex` is back as the fast default.

Reply:
> Now that that `substr` copy is an rvalue, I believe C++'s copy elision will take care of eliminating temporaries better.

Yes, but `std::string::substr()` will still result in the creation of a `std::string` object whereas `std::string_view::substr()` will put a range on the existing allocation for the `RegexMatch()` parameter. We cannot get rid of the new object inside that call.

Reply:
As you note, I guess no matter what, a fresh string is required for both `llvm::Regex` (for anchoring) and `std::regex` (because the constructor takes a `std::string` and not a `std::string_view`).

So I'm not sure what you're suggesting...?

At any rate, it would be nice to get this regex machinery merged. Any optimizations on top of that would be welcome if they make a significant improvement for `std::regex`.

Reply:
> As you note, I guess no matter what, a fresh string is required for both `llvm::Regex` (for anchoring) and `std::regex` (because the constructor takes a `std::string` and not a `std::string_view`).
> 
> So I'm not sure what you're suggesting...?
> 
> At any rate, it would be nice to get this regex machinery merged. Any optimizations on top of that would be welcome if they make a significant improvement for `std::regex`.

Yes. It's micro-optimizations not necessary for this to land. But while you are at it. If I would build it myself I could just test it myself and post a PR but I am not gonna burden myself with another project right now. Still got way too much backlog. 😓

Go ahead and merge this.

Reply:
I tried some half-hearted attempt at having `ExtractKeysMarkedAsRegexes` return a vector of `std::string_view`, but the after-effects caused even more temporaries for the subsequent map lookups and regex eval. And I couldn't see any impact on execution time one way or the other.

So I think reducing temporaries here is a laudable goal, but it's probably not super-trivial.

Thanks for the review!